Remove support for node.client setting

javanna opened the pull request:
As discussed in #16565, the node.client setting is an unnecessary shortcut for `node.data: false` and `node.master: false`. We have places where we treat nodes with `node.client` set to `true` differently compared to when nodes have both `node.master` set to `false` and `node.data` set to `false`. Also, with the addition of `node.ingest` (or potentially new roles in the future), it becomes confusing to figure out whether a node client will act as ingest node or not. 

This PR removes the `node.client` setting in favour of explicitly using `node.master`, `node.data` and `node.ingest` instead. Note that `node.client` is now a prohibited node setting rather than just being ignored, so that users know they have to upgrade their settings. A node will not start if it has `node.client` among its settings.

Additionally, tribe nodes and all of the tribe clients are now coordinating nodes only (with the addition of the ingest node type they mistakenly became ingest nodes as well, which they shouldn't have done).

Internally, the DiscoveryNode class supports now the notion of roles, through the inner Role enum. A node with no roles is implicitly a coordinating only node, which is the least that a node can do.

The following are the breaking changes made on the REST (and java api) layer:
- `_cat/nodes` used to return `c` for client node or `d` for data node as part of the `node.role` column. It now returns `m` for master eligible, `d` for data and/or `i` for ingest. A node with no explicit roles will be a coordinating only node and marked with `-`. A node can have multiple roles. The master column has been adapted to return only whether a node is the current master (`*`) or not (`-`).

- The cluster stats api now returns counts for each node role. The `master_data`, `master_only`, `data_only` and `client` fields have been removed from the response in favour of `master`, `data`, `ingest` and `coordinating_only`. The same node can have multiple roles, thus contributing to multiple role counts. Every node is implicitly a coordinating node, so whenever a node has no explicit roles, it will be counted as coordinating only.

In addition, the documentation has been adapted and some tests renamed in our test infra.

Node roles and node attributes have been separated internally within `DiscoveryNode`. Previously node roles would get printed out as part of node attributes in nodes info, nodes stats and tasks list apis, but now that they are stored separately they are printed out as part of a new field of type array called `roles`. One nice side effect of this is that roles will contain also the default ones, while when printed as part of attributes or settings only the ones that differ from default would get printed.

Closes #16565

javanna commented:
I merged master in after two weeks of inactivity and make everything work again, reviews are welcome!

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/liveness/LivenessResponse.java:
```diff
@@ -59,7 +59,12 @@ public void readFrom(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         clusterName.writeTo(out);
-        out.writeOptionalStreamable(node);
```
use `out.writeOptionalWriteable`

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:
```diff
@@ -213,25 +215,29 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        private int total;
+        private Map<String, Integer> roles;
+
+        Counts() {
```
any chance we can make this writeable and have all members being final?

s1monw commented on core/src/main/java/org/elasticsearch/client/transport/TransportClient.java:
```diff
@@ -112,7 +112,10 @@ private PluginsService newPluginService(final Settings settings) {
                     .put(NetworkService.NETWORK_SERVER.getKey(), false)
-                    .put(Node.NODE_CLIENT_SETTING.getKey(), true)
+                    //nocommit not too sure if these settings are needed here...
+                    .put(Node.NODE_MASTER_SETTING.getKey(), false)
```
why do we need this, we know that we are a transport client, can't we make this explicit somehow?

javanna replied:
yea I think we already do through client_type setting below.... was wondering the same, hence the nocommit...

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -338,25 +336,24 @@ public Version getVersion() {
-        DiscoveryNode node = new DiscoveryNode();
-        node.readFrom(in);
-        return node;
+        return PROTOTYPE.readFrom(in);
```
can we get rid of `PROTOTYPE ` and remove the readNode method alltogether. I thikn we should just have a `DiscoveryNode(StreamInput in)` method and do all the reading there.

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -193,20 +170,58 @@ public DiscoveryNode(String nodeName, String nodeId, String hostName, String hos
+            this.version = version;
+        }
+        this.attributes = attributesBuilder.build();
+        this.roles = resolveRoles(this.attributes);
```
can we make a `Collections.immutableSet` here?

javanna replied:
done below in resolveRoles

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeService.java:
```diff
@@ -62,19 +63,21 @@ public DiscoveryNodeService addCustomAttributeProvider(CustomAttributesProvider
-            if (attributes.get("data").equals("true")) {
-                attributes.remove("data");
+        //nocommit why don't we remove master as well if it's true? and ingest?
+        if (attributes.containsKey(DiscoveryNode.Role.DATA.getRoleName())) {
```
dude this entire method makes no sense to me. Can we just make sure we make these attributes dense rather than sparse and just fill all the roles for every node?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -376,18 +376,24 @@ public DiscoveryNode resolveNode(String node) {
                         String matchAttrName = nodeId.substring(0, index);
                         String matchAttrValue = nodeId.substring(index + 1);
-                        if ("data".equals(matchAttrName)) {
+                        if (DiscoveryNode.Role.DATA.getRoleName().equals(matchAttrName)) {
```
maybe use swtich here?

javanna replied:
won't work as we call getRoleName(): constant expression required.

s1monw replied:
got it

s1monw commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -849,10 +849,11 @@ private DiscoveryNode findMaster() {
-                // filter out the client node, which is a client node, or also one that is not data and not master (effectively, client)
-            } else if (masterElectionFilterDataNodes && (!node.masterNode() && node.dataNode())) {
-                // filter out data node that is not also master
+            //nocommit we should rename this and its setting, also we ignore node.ingest, but maybe it's ok here
```
not sure what you mean by that nocommit?

javanna replied:
I mean that we remove any mention of client nodes but we still have a setting called masterElectionFilterClientNodes. Not sure we should address this here though.

s1monw replied:
make it a TODO?

javanna replied:
sure, also opened #17325 so we don't forget

s1monw commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -143,7 +143,7 @@
         Setting.boolSetting("node.ingest", true, Property.NodeScope);
     public static final Setting<String> NODE_NAME_SETTING = Setting.simpleString("node.name", Property.NodeScope);
-    // this sucks that folks can mistype client etc and get away with it.
+    // this sucks that folks can mistype data, master or ingest and get away with it.
```
I wonder if we should fix that at the same go and introduce `node.attributes.${someuserattr}` and make it strict? How much easier would your live become here?

javanna replied:
I don;t think this would have made this specific PR easier, it would remove a weird if that's it. I would do that as a follow-up, don't want this PR to get crazy big.

s1monw replied:
sure no worries, I will open an issue for this anyway

javanna replied:
for reference, here is the issue that Simon opened: #17280

s1monw commented on core/src/main/java/org/elasticsearch/rest/action/cat/RestNodesAction.java:
```diff
@@ -279,8 +280,15 @@ private Table buildTable(RestRequest req, ClusterStateResponse state, NodesInfoR
-            table.addCell(masterId == null ? "x" : masterId.equals(node.id()) ? "*" : node.masterNode() ? "m" : "-");
+
+            String roles;
+            if (node.getRoles().size() == 0) {
```
`if (node.getRolse().isEmpty())`

s1monw replied:
roles can be final?

s1monw commented on core/src/main/java/org/elasticsearch/transport/ConnectTransportException.java:
```diff
@@ -61,7 +61,12 @@ public ConnectTransportException(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeOptionalStreamable(node);
```
optional writeable?

s1monw commented:
@javanna I left some suggestiosn and questions

s1monw commented:
@javanna what's the status here I'd still like to see [this](https://github.com/elastic/elasticsearch/pull/16963/files#r57137919) addressed

javanna commented:
@s1monw I am working on it it's the last one on my list, I haven't said I addressed all the comments yet :) will ping you when ready for another review

javanna commented:
I pushed a bunch of new commits and merged master a number of times. It's much better now, all the nocommits are gone and attributes and roles are well separated within `DiscoveryNode`. I added a new `roles` field in the output of nodes stats, nodes info and tasks list api, because we would previously print node types as part of attributes when they differ from the default. The main reason why the PR got so much bigger is the move of some Streamable to Writeable and the fact that we now require to specify attributes and roles when creating a `DiscoveryNode` instance (I didn't want to have an empty default nor a default with all the roles, better be explicit). This last one is a test problem only, as DiscoveryNodeService takes care of creating nodes in production and does the right thing. Also fixed some line length problems while touching files.

@s1monw ready for another round of review ;)

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoResponse.java:
```diff
@@ -84,6 +86,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 }
             }
 
+            builder.array("roles", nodeInfo.getNode().getRoles().stream().map(DiscoveryNode.Role::getRoleName)
```
can we just use a for loop here, I think we should not mix streams in too much :)

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:
```diff
@@ -41,40 +42,57 @@
+
+        size = in.readVInt();
+        this.plugins = new HashSet<>(size);
+        for (; size > 0; size--) {
```
maybe remove the optimization and just use an ordinary incrementing for loop?

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/stats/ClusterStatsNodes.java:
```diff
@@ -479,20 +466,54 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        private JvmStats(StreamInput in) throws IOException {
+            int size = in.readVInt();
+            this.versions = new ObjectIntHashMap<>(size);
+            for (; size > 0; size--) {
```
oooh another one :)

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -96,39 +82,64 @@ public static boolean ingestNode(Settings settings) {
+    private final String hostName;
+    private final String hostAddress;
+    private final TransportAddress address;
+    private final ImmutableOpenMap<String, String> attributes;
```
I wonder if we really need this ImmutableOpenMap or if we can just use a default one from the JDK?

javanna replied:
agreed I am getting rid of it

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -158,55 +171,73 @@ public DiscoveryNode(String nodeName, String nodeId, TransportAddress address, M
+        this.attributes = attributesBuilder.build();
+        //verify that no node roles are being provided as attributes
+        boolean assertEnabled = false;
+        assert assertEnabled = true;
```
the way I do that is:

```Java

assert () -> { for (Role role : Role.values()) {
                 assert attributes.containsKey(role.getRoleName()) == false;
            }; return true;}
```

and please add a msg to the assertion like:

```Java
assert attributes.containsKey(role.getRoleName()) == false : "role name: " + role.getRoleName() + " found in attributes";
```

javanna replied:
I added the message....but the function magic didn't work for me...do we have that anywhere in our codebase already?

s1monw replied:
```
Predicate<Map<String, String>> predicate =  (attrs) -> {
    for (Role role : Role.values()) {
        assert attrs.containsKey(role.getRoleName()) == false;
    }
    return true;
};
assert predicate.test(attributes);
```

this should work

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -158,55 +171,73 @@ public DiscoveryNode(String nodeName, String nodeId, TransportAddress address, M
+    }
+
+    private static ImmutableOpenMap.Builder<String, String> copyAttributes(ImmutableOpenMap<String, String> attributes) {
+        ImmutableOpenMap.Builder<String, String> builder = ImmutableOpenMap.builder();
```
oh boy this string interning has to stop here.. not this change....

s1monw replied:
oh wait, we now have roles, I think in this case we can just drop the attributes interning?

javanna replied:
yes I think so

s1monw replied:
thanks!!!

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -317,8 +328,11 @@ public boolean isMasterNode() {
+        return roles.contains(Role.INGEST);
+    }
+
+    public Set<Role> getRoles() {
```
can you javadoc this! I really like that part of the change

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -96,39 +82,64 @@ public static boolean ingestNode(Settings settings) {
+        }
+        this.attributes = attributesBuilder.build();
+        int rolesSize = in.readVInt();
+        this.roles = new HashSet<>(rolesSize);
```
Also I think this should be an `EnumSet` and maybe just read using a switch statement?

javanna replied:
moved to EnumSet, not sure about the switch, do you want to move away from using ordinals for serialization?

s1monw replied:
nevermind not that important

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -425,4 +426,26 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
+
+    public enum Role {
```
maybe add some javadocs

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeService.java:
```diff
@@ -62,19 +67,20 @@ public DiscoveryNodeService addCustomAttributeProvider(CustomAttributesProvider
-                // if we are client node, don't store data ...
-                attributes.put("data", "false");
-            }
+            throw new IllegalArgumentException("node.client setting is no longer supported, use " + Node.NODE_MASTER_SETTING.getKey()
```
oh this is only possible because we allos `"node.*"` as a valid setting... grrr :)

javanna replied:
yea....next step! :)

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeService.java:
```diff
@@ -92,9 +98,8 @@ public DiscoveryNode buildLocalNode(TransportAddress publishAddress) {
                 logger.warn("failed to build custom attributes from provider [{}]", e, provider);
```
`CustomAttributesProvider` gave me a hard-attack, can you open an issue to move the registration to a module please?

javanna replied:
Opened #17377

s1monw commented:
@javanna we are close, I left another round of review comments

javanna commented:
Ready for another round of review, left a couple of questions on review comments. @clintongormley maybe you want to have a look at the api changes in nodes info, nodes stats, cluster stats, cat nodes and tasks list? They are all listed in the description of the PR.

s1monw commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -366,8 +365,8 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
                     Map<String, String> tribeAttr = new HashMap<>();
-                    for (ObjectObjectCursor<String, String> attr : tribe.attributes()) {
-                        tribeAttr.put(attr.key, attr.value);
+                    for (Map.Entry<String, String> entry : tribe.getAttributes().entrySet()) {
```
can we use `new HashMap(tribe.getAttributes());`

javanna replied:
oops good catch

s1monw commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:
```diff
@@ -218,7 +218,9 @@ public DiscoveryNode(String nodeName, String nodeId, String hostName, String hos
             }
         }
-        this.roles = roles;
+        Set<Role> rolesSet = Collections.unmodifiableSet(roles);
```
I think this `Collections.unmodifiableSet(roles);` is unnecessary and we can just do `this.roles.addAll(roles)` ?

s1monw commented:
@javanna I replied to your comments and added some myself. I think we don't need another round of review - if you feel you addressed it go push... LGTM

jaymode commented on core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java:
```diff
@@ -378,11 +380,13 @@ public LivenessResponse newInstance() {
+                        // so people can control which address is exactly used.
                         DiscoveryNode nodeWithInfo = livenessResponse.getDiscoveryNode();
-                        newNodes.add(new DiscoveryNode(nodeWithInfo.name(), nodeWithInfo.id(), nodeWithInfo.getHostName(), nodeWithInfo.getHostAddress(), listedNode.address(), nodeWithInfo.attributes(), nodeWithInfo.version()));
+                        newNodes.add(new DiscoveryNode(nodeWithInfo.name(), nodeWithInfo.id(), nodeWithInfo));
```
I think we lost the `listedNode.address()` storage with this change

javanna replied:
yes we did, thanks a lot for catching it, I fixed and added a test for this. Mind having a look at it? I don't personally love it but couldn't find a better way to test it...

jaymode commented:
@javanna the fix in the TransportClientNodesService LGTM
